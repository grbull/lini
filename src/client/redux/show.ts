/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { ShowWithEpisodesDto } from '../../server/show/show.dto';
import { api } from '../utils/api';

type ShowState = {
  status: 'init' | 'loading' | 'idle' | 'error';
  data?: ShowWithEpisodesDto;
  error?: string;
};

const initialState: ShowState = {
  status: 'init',
};

export const get = createAsyncThunk<ShowWithEpisodesDto, number>(
  'show/get',
  (id) => api.show.get(id)
);

export const show = createSlice({
  name: 'show',
  initialState,
  reducers: {
    reset: () => {
      return initialState;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(get.pending, (state) => {
      state.status = 'loading';
      state.data = undefined;
      state.error = undefined;
    });
    builder.addCase(get.fulfilled, (state, { payload }) => {
      state.status = 'idle';
      state.data = payload;
    });
    builder.addCase(get.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });
  },
});

export const showActions = {
  ...show.actions,
  get,
};
