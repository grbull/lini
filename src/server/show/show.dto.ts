import { EpisodeDto } from '../episode/episode.dto';
import { NetworkDto } from '../network/network.dto';
import { DayOfWeek } from '../tvmaze/tvmaze.types';
import { WebChannelDto } from '../web_channel/web_channel.dto';

export class ShowDto {
  id!: number;
  name!: string;
  type!: string;
  language?: string;
  genres!: string[];
  status!: string;
  runtime!: number | null;
  runtimeAverage?: number;
  datePremiered?: string;
  officialSite?: string;
  scheduleTime!: string | null;
  scheduleDays!: (keyof typeof DayOfWeek)[];
  rating?: number;
  network?: NetworkDto;
  webChannel?: WebChannelDto;
  imageMedium?: string;
  imageOriginal?: string;
  summary?: string;
}

export class ShowWithEpisodesDto extends ShowDto {
  episodes!: EpisodeDto[];
}
