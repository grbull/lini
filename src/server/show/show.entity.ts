import { Exclude } from 'class-transformer';
import {
  Column,
  CreateDateColumn,
  Entity,
  ManyToOne,
  OneToMany,
  UpdateDateColumn,
} from 'typeorm';

import { EpisodeEntity } from '../episode/episode.entity';
import { NetworkEntity } from '../network/network.entity';
import { SubscriptionEntity } from '../subscription/subscription.entity';
import { DayOfWeek } from '../tvmaze/tvmaze.types';
import { WebChannelEntity } from '../web_channel/web_channel.entity';

@Entity('show')
export class ShowEntity {
  @Column('integer', { primary: true })
  id!: number;

  @Column('text')
  name!: string;

  // Scripted/Documentary/Reality/etc
  @Column('text')
  type!: string;

  @Column('text', { nullable: true })
  language!: string | null;

  @Column('text', { array: true })
  genres!: string[];

  // "Running or ended"
  @Column('text')
  status!: string;

  @Column('integer', { nullable: true })
  runtime!: number | null;

  @Column('integer', { nullable: true })
  runtimeAverage!: number | null;

  @Column('date', { nullable: true })
  datePremiered!: string | null;

  @Column('text', { nullable: true })
  officialSite!: string | null;

  @Column('time', { nullable: true })
  scheduleTime!: string | null;

  @Column('enum', { enum: DayOfWeek, array: true })
  scheduleDays!: (keyof typeof DayOfWeek)[];

  @Column('float', { nullable: true })
  rating!: number | null;

  @Column('integer')
  @Exclude()
  weight!: number;

  @ManyToOne(() => NetworkEntity, (network) => network.id, {
    nullable: true,
  })
  network!: NetworkEntity | null;

  @ManyToOne(() => WebChannelEntity, (webChannel) => webChannel.id, {
    nullable: true,
  })
  webChannel!: WebChannelEntity | null;

  @Column('integer', { nullable: true })
  @Exclude()
  tvrage!: number | null;

  @Column('integer', { nullable: true })
  @Exclude()
  thetvdb!: number | null;

  @Column('text', { nullable: true })
  @Exclude()
  imdb!: string | null;

  @Column('text', { nullable: true })
  imageMedium!: string | null;

  @Column('text', { nullable: true })
  imageOriginal!: string | null;

  @Column('text', { nullable: true })
  summary!: string | null;

  @OneToMany(() => EpisodeEntity, (episode) => episode.show)
  episodes!: EpisodeEntity[];

  @Column('bool', { default: false, select: false })
  @Exclude()
  isSeeded!: boolean;

  @OneToMany(() => SubscriptionEntity, (subscription) => subscription.show)
  @Exclude()
  subscriptions!: SubscriptionEntity[];

  @CreateDateColumn({ type: 'timestamptz' })
  @Exclude()
  dateCreated!: string;

  @UpdateDateColumn({ type: 'timestamptz' })
  @Exclude()
  dateUpdated!: string;
}
