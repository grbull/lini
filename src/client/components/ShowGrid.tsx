import cn from 'classnames';
import React, { ReactElement } from 'react';
import { Link } from 'react-router-dom';

import { ShowDto } from '../../server/show/show.dto';
import { imageShow } from '../utils/imageShow';

interface Props {
  shows: ShowDto[];
}

// TODO: Min height, so things don't change very much when images are loading
// I assume it needs to be % for responisive
// Problem making the title/rating div use all available space.

export function ShowGrid({ shows }: Props): ReactElement {
  return (
    <div className={cn('flex', 'flex-row', 'flex-wrap', 'px-2')}>
      {shows.map((show) => (
        <div className={cn('w-1/3', 'p-1')} key={show.id}>
          <Link
            className={cn(
              'flex',
              'flex-col',
              'bg-gray-200 dark:bg-gray-800',
              'h-full',
              'border-b-2',
              'border-solid',
              { 'border-green-500': show.status === 'Running' },
              {
                'border-yellow-500':
                  show.status === 'To Be Determined' ||
                  show.status === 'In Development',
              },
              { 'border-red-500': show.status === 'Ended' }
            )}
            to={`/show/${show.id}`}
          >
            <img
              alt={show.name}
              className={cn('w-full')}
              src={imageShow(show.imageMedium)}
            />
            <div
              className={cn(
                'min-h-16',
                'flex-grow',
                'flex',
                'flex-col',
                'justify-around',
                'items-center',
                'p-2'
              )}
            >
              <p className={cn('text-center')}>{show.name}</p>
            </div>
          </Link>
        </div>
      ))}
    </div>
  );
}
