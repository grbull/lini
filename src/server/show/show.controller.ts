import {
  Controller,
  Get,
  Param,
  Query,
  UseGuards,
  UseInterceptors,
} from '@nestjs/common';

import { AuthGuard } from '../common/guards/auth.guard';
import { LoggingInterceptor } from '../common/interceptors/logging.interceptor';
import { ShowEntity } from './show.entity';
import { ShowService } from './show.service';

@Controller('show')
@UseGuards(AuthGuard)
@UseInterceptors(LoggingInterceptor)
export class ShowController {
  constructor(private readonly showService: ShowService) {}

  @Get('autocomplete')
  public autoComplete(@Query('q') query: string): Promise<ShowEntity[]> {
    return this.showService.autocomplete(query);
  }

  // TODO: Implement pagination
  @Get('search/:searchQuery')
  public search(
    @Param('searchQuery') searchQuery: string
  ): Promise<ShowEntity[]> {
    return this.showService.search(searchQuery);
  }

  @Get(':id')
  public getOne(@Param('id') id: number): Promise<ShowEntity> {
    return this.showService.getOne(id);
  }

  // public getShowsByLanguage() {}

  // public getShowsByNetwork() {}

  // public getShowsByWebChannel() {}
}
