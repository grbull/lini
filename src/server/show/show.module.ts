import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';

import { LoggerModule } from '../logger/logger.module';
import { NetworkModule } from '../network/network.module';
import { WebChannelModule } from '../web_channel/web_channel.module';
import { ShowController } from './show.controller';
import { ShowEntity } from './show.entity';
import { ShowService } from './show.service';

@Module({
  imports: [
    LoggerModule,
    NetworkModule,
    WebChannelModule,
    TypeOrmModule.forFeature([ShowEntity]),
  ],
  providers: [ShowService],
  controllers: [ShowController],
  exports: [ShowService],
})
export class ShowModule {}
