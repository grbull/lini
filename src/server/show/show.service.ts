import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Raw, Repository } from 'typeorm';

import { LoggerService } from '../logger/logger.service';
import { NetworkService } from '../network/network.service';
import { TvMazeShow, TvMazeShowEmbedded } from '../tvmaze/tvmaze.types';
import { TvMazeShowToEntity } from '../tvmaze/tvmaze.utils';
import { WebChannelService } from '../web_channel/web_channel.service';
import { ShowEntity } from './show.entity';

@Injectable()
export class ShowService {
  constructor(
    private readonly loggerService: LoggerService,
    private readonly networkService: NetworkService,
    private readonly webChannelService: WebChannelService,
    @InjectRepository(ShowEntity)
    private readonly showRepository: Repository<ShowEntity>
  ) {
    this.loggerService.setContext('ShowService');
  }

  /**
   * Used for search and autocomplete when typing in a search box. Returns 5 items
   * Considering returning paginated result for search. TBD
   */
  public autocomplete(query: string): Promise<ShowEntity[]> {
    return this.showRepository
      .createQueryBuilder('show')
      .where({
        name: Raw((name) => `LOWER(${name}) LIKE '%${query.toLowerCase()}%'`),
      })
      .leftJoinAndSelect('show.network', 'network')
      .leftJoinAndSelect('network.country', 'networkCountry')
      .leftJoinAndSelect('show.webChannel', 'webChannel')
      .leftJoinAndSelect('webChannel.country', 'webChannelCountry')
      .take(5)
      .getMany();
  }

  public search(query: string): Promise<ShowEntity[]> {
    return this.showRepository
      .createQueryBuilder('show')
      .where({
        name: Raw((name) => `LOWER(${name}) LIKE '%${query.toLowerCase()}%'`),
      })
      .leftJoinAndSelect('show.network', 'network')
      .leftJoinAndSelect('network.country', 'networkCountry')
      .leftJoinAndSelect('show.webChannel', 'webChannel')
      .leftJoinAndSelect('webChannel.country', 'webChannelCountry')
      .take(48)
      .getMany();
  }

  public getOne(id: number): Promise<ShowEntity> {
    try {
      return this.showRepository
        .createQueryBuilder('show')
        .where('show.id = :id', { id })
        .leftJoinAndSelect('show.network', 'network')
        .leftJoinAndSelect('network.country', 'networkCountry')
        .leftJoinAndSelect('show.webChannel', 'webChannel')
        .leftJoinAndSelect('webChannel.country', 'webChannelCountry')
        .leftJoinAndSelect('show.episodes', 'episode')
        .orderBy('episode.airdate', 'ASC')
        .getOneOrFail();
    } catch {
      throw new NotFoundException();
    }
  }

  // All of the below is used primarily for seeding and updating from tvmaze
  private async createOneFromTvMaze(
    show: TvMazeShow | TvMazeShowEmbedded
  ): Promise<ShowEntity> {
    const webChannel = show.webChannel
      ? (await this.webChannelService.createOrUpdate(show.webChannel)).id
      : null;
    const network = show.network
      ? (await this.networkService.createOrUpdate(show.network)).id
      : null;

    return this.showRepository.create(
      TvMazeShowToEntity(show, network, webChannel)
    );
  }

  private async createManyFromTvMaze(
    shows: TvMazeShow[] | TvMazeShowEmbedded[]
  ): Promise<ShowEntity[]> {
    const showEntities: ShowEntity[] = [];

    // Must be done synchronously to avoid duplicate relations
    for (const show of shows) {
      showEntities.push(await this.createOneFromTvMaze(show));
    }
    return showEntities;
  }

  private async filterExisting(shows: TvMazeShow[]): Promise<TvMazeShow[]> {
    const existingShows = await this.showRepository.find({
      select: ['id'],
      where: shows.map((show) => Object({ id: show.id })),
    });

    return shows.filter(
      (show) =>
        !existingShows.find((existingShow) => existingShow.id === show.id)
    );
  }

  /**
   * Used when seeding series. We filter out existing to avoid altering
   * the update column. We want to avoid marking series as updated, when we
   * havn't updated the relevant episodes.
   */
  public async createManyIfDoesNotExist(
    shows: TvMazeShow[]
  ): Promise<ShowEntity[]> {
    const showsToCreate = await this.filterExisting(shows);
    const showEntities = await this.createManyFromTvMaze(showsToCreate);

    return this.showRepository.save(showEntities);
  }

  public async deleteOneById(showId: number): Promise<void> {
    await this.showRepository.delete(showId);
  }

  public async updateShow(show: TvMazeShowEmbedded): Promise<ShowEntity> {
    const showEntity = await this.createOneFromTvMaze(show);
    return this.showRepository.save(showEntity);
  }

  public async markSeeded(showId: number): Promise<void> {
    await this.showRepository.update(showId, { isSeeded: true });
  }

  public getShowsToSeed(): Promise<
    Pick<ShowEntity, 'id' | 'isSeeded' | 'name'>[]
  > {
    return this.showRepository
      .createQueryBuilder('show')
      .select(['show.id', 'show.isSeeded', 'show.name'])
      .where('show.isSeeded = :isSeeded', { isSeeded: false })
      .orderBy('show.id', 'ASC')
      .getMany();
  }

  public getShowsLastUpdated(): Promise<
    Pick<ShowEntity, 'id' | 'dateUpdated'>[]
  > {
    return this.showRepository
      .createQueryBuilder('show')
      .select(['show.id', 'show.dateUpdated'])
      .orderBy('show.id', 'ASC')
      .getMany();
  }
}
