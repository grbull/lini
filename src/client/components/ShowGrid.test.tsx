/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom/extend-expect';

import { render, RenderResult } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import React, { ReactNode } from 'react';
import { Router } from 'react-router-dom';

import { ShowDto } from '../../server/show/show.dto';
import { ShowGrid } from './ShowGrid';

const showDto: ShowDto = {
  id: 216,
  name: 'Rick and Morty',
  type: 'Animation',
  language: 'English',
  genres: ['Comedy', 'Adventure', 'Science-Fiction'],
  status: 'Running',
  runtime: 30,
  runtimeAverage: 30,
  datePremiered: '2013-12-02',
  officialSite: 'http://www.adultswim.com/videos/rick-and-morty',
  scheduleTime: '23:00:00',
  scheduleDays: ['Sunday'],
  rating: 9,
  imageMedium: '/uploads/images/medium_portrait/1/3603.jpg',
  imageOriginal: '/uploads/images/original_untouched/1/3603.jpg',
  summary:
    'Rick is a mentally gifted, but sociopathic and alcoholic scientist and a grandfather to Morty; an awkward, impressionable, and somewhat spineless teenage boy. Rick moves into the family home of Morty, where he immediately becomes a bad influence.',
  network: {
    id: 10,
    name: 'Adult Swim',
    country: {
      name: 'United States',
      code: 'US',
      timezone: 'America/New_York',
    },
  },
  webChannel: null,
};

function setup(children: ReactNode): RenderResult {
  const history = createMemoryHistory();

  const utils = render(<Router history={history}>{children}</Router>);

  return { ...utils };
}

describe('ShowGrid Component', () => {
  it('matches the snapshot', () => {
    const { asFragment } = setup(<ShowGrid shows={[showDto]} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches the snapshot when empty', () => {
    const { asFragment } = setup(<ShowGrid shows={[]} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
