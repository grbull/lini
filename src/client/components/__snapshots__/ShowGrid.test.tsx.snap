// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowGrid Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-row flex-wrap px-2"
  >
    <div
      class="w-1/3 sm:w-1/4 md:w-1/5 p-1"
    >
      <a
        class="flex flex-col bg-gray-200 dark:bg-gray-800 h-full border-b-2 border-solid border-green-500 shadow"
        href="/show/216"
      >
        <div
          class="h-0 relative box-border p-1 border-2 border-gray-800"
          style="padding-top: 140.47619047619045%;"
        >
          <div
            class="animate-pulse bg-gray-700 absolute top-0 left-0 w-full h-full bg-gray-800"
          />
          <img
            alt="Rick and Morty"
            class="absolute top-0 left-0 w-full h-full hidden"
            src="undefined/uploads/images/medium_portrait/1/3603.jpg"
          />
        </div>
        <div
          class="min-h-16 flex-grow flex flex-col justify-around items-center p-2"
        >
          <p
            class="text-center"
          >
            Rick and Morty
          </p>
        </div>
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ShowGrid Component matches the snapshot when empty 1`] = `
<DocumentFragment>
  <p
    class="px-2"
  >
    No results to display.
  </p>
</DocumentFragment>
`;
